client/exec.go

```diff
@@ -0,0 +1,312 @@
+	Stderr io.Writer
+}
+
+type execPayload struct {
+	Command       []string          `json:"command"`
+	Environment   map[string]string `json:"environment"`
+	WorkingDir    string            `json:"working-dir"`
+	Timeout       string            `json:"timeout"`
+	UserID        *int              `json:"user-id"`
+	User          string            `json:"user"`
+	GroupID       *int              `json:"group-id"`
+	Group         string            `json:"group"`
+	UseTerminal   bool              `json:"use-terminal"`
+	CombineStderr bool              `json:"combine-stderr"`
+	Width         int               `json:"width"`
+	Height        int               `json:"height"`
+}
+
+type execResult struct {
+	WebsocketIDs map[string]string `json:"websocket-ids"`
+}
+
+// ExecProcess represents a running process. Use Wait to wait for it to finish.
+type ExecProcess struct {
+	changeID    string
+	client      *Client
+	timeout     time.Duration
+	writesDone  chan struct{}
+	controlConn jsonWriter
+}
+
+// jsonWriter makes it easier to write tests for SendSignal and SendResize.
+type jsonWriter interface {
+	WriteJSON(v interface{}) error
+}
+
+// Exec starts a command with the given options, returning a value
+// representing the process.
+func (client *Client) Exec(opts *ExecOptions) (*ExecProcess, error) {
+	// Set up stdin/stdout/stderr defaults.
+	stdin := opts.Stdin
+	if stdin == nil {
+		stdin = bytes.NewReader(nil)
+	}
+	stdout := opts.Stdout
+	if stdout == nil {
+		stdout = ioutil.Discard
+	}
+	var stderr io.Writer
+	if !opts.CombineStderr {
+		stderr = opts.Stderr
+		if stderr == nil {
+			stderr = ioutil.Discard
+		}
+	} else if stderr != nil {
+		return nil, fmt.Errorf("opts.Stderr must be nil if opts.CombineStderr is true")
+	}
+
+	// Call the /v1/exec endpoint to start the command.
+	var timeoutStr string
+	if opts.Timeout != 0 {
+		timeoutStr = opts.Timeout.String()
+	}
+	payload := execPayload{
+		Command:       opts.Command,
+		Environment:   opts.Environment,
+		WorkingDir:    opts.WorkingDir,
+		Timeout:       timeoutStr,
+		UserID:        opts.UserID,
+		User:          opts.User,
+		GroupID:       opts.GroupID,
+		Group:         opts.Group,
+		UseTerminal:   opts.UseTerminal,
+		CombineStderr: opts.CombineStderr,
+		Width:         opts.Width,
+		Height:        opts.Height,
+	}
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(&payload)
+	if err != nil {
+		return nil, err
+	}
+	headers := map[string]string{
+		"Content-Type": "application/json",
+	}
+	resultBytes, changeID, err := client.doAsyncFull("POST", "/v1/exec", nil, headers, &body)
+	if err != nil {
+		return nil, err
+	}
+	var result execResult
+	err = json.Unmarshal(resultBytes, &result)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check that we have the websocket IDs we expect.
+	wsIDs := result.WebsocketIDs
+	if wsIDs["control"] == "" {
+		return nil, fmt.Errorf(`response did not include "control" websocket`)
```

Comment:
Suggested wording, with a bit more context (exec), and rephrased for terseness:

`exec response lacks "control" websocket`
`exec response lacks "io" websocket`
`exec response lacks "stderr" websocket`

Reply:
Done. Agreed on more context and rephrasing for terseness, though I used "missing" instead of "lacks" as it seemed more natural.